server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java

```diff
@@ -264,19 +293,45 @@ public void update(TableSchemaDesc schema, List<Map<String, Object>> records) {
         this.lastUpdateTime = System.currentTimeMillis();
         this.schema = newSchema;
         if (decodedRecords != null) {
-            this.insertRecords(decodedRecords);
+            this.insertRecords(timestamp, decodedRecords);
         }
     }
 
-    private void insertRecords(List<Map<String, Object>> records) {
+    private void insertRecords(long timestamp, List<Map<String, Object>> records) {
+        var typeMap = this.schema.getColumnTypeMapping();
         for (var record : records) {
-            var key = record.get(this.schema.getKeyColumn());
-            if (record.get("-") != null) {
-                this.recordMap.remove(key);
+            var newRecord = new HashMap<>(record);
+            var key = newRecord.remove(this.schema.getKeyColumn());
+            var deletedFlag = (Boolean) newRecord.remove(DELETED_FLAG_COLUMN_NAME);
+            if (deletedFlag == null) {
+                deletedFlag = false;
+            }
+            var versions = this.recordMap.computeIfAbsent(key, k -> new ArrayList<>());
+            if (deletedFlag) {
+                if (versions.isEmpty() || !versions.get(versions.size() - 1).isDeleted()) {
+                    versions.add(MemoryRecord.builder()
+                            .timestamp(timestamp)
+                            .deleted(true)
+                            .build());
+                }
```

Comment:
Can this block be removed? 
It seems that it's just a special condition of the logic below.

Reply:
Their conditions are different: "!versions.get(versions.size() - 1).isDeleted()" vs "!newRecord.isEmpty()", so it is better to put them in two clauses.